src/widgets/taskbarnotifier.cpp

```diff
@@ -0,0 +1,251 @@
+#endif
+}
+
+#elif defined(Q_OS_WINDOWS)
+QImage TaskBarNotifier::Private::makeIconCaption(const QImage &image, const QString &number) const
+{
+    if (!image.isNull()) {
+        auto   imSize    = image.size() * devicePixelRatio_;
+        QImage img       = image;
+        auto   letters   = number.length();
+        auto   text      = QStaticText((letters < 3) ? number : "99+");
+        auto   textDelta = (letters <= 2) ? 3 : 4;
+
+        QPainter p(&img);
+        p.setRenderHints(QPainter::Antialiasing | QPainter::TextAntialiasing);
+        auto font   = QFont("Times", imSize.height() / textDelta, QFont::Bold);
+        auto fm     = QFontMetrics(font);
+        auto fh     = fm.height();
+        auto fw     = fm.horizontalAdvance(text.text());
+        auto radius = fh / 2;
+
+        Qt::BrushStyle style = Qt::SolidPattern;
+        QBrush         brush(Qt::red, style);
+        p.setBrush(brush);
+        p.setPen(QPen(Qt::NoPen));
+        QRect rect(imSize.width() - fw - radius, 0, fw + radius, fh);
+        p.drawRoundedRect(rect, radius, radius);
+
+        p.setFont(font);
+        p.setPen(QPen(Qt::white));
+        auto offset = rect.width() / ((letters + 1) * 2);
+        p.drawStaticText(rect.x() + offset, rect.y(), text);
+
+        p.end();
+        return img;
+    }
+    return QImage();
+}
+
+void TaskBarNotifier::Private::setParentIcon(const QImage &image, uint count)
+{
+    if (image.isNull())
+        return;
+
+    QImage img;
+    if (count > 0 && urgent_) {
+        img = makeIconCaption(image, QString::number(count));
+    }
+    HICON icon = (img.isNull()) ? getHICONfromQImage(image) : getHICONfromQImage(img);
+    SendMessage(hwnd_, WM_SETICON, ICON_SMALL, (LPARAM)icon);
+    SendMessage(hwnd_, WM_SETICON, ICON_BIG, (LPARAM)icon);
+    doFlashTaskbarIcon();
+}
+
+HICON TaskBarNotifier::Private::getHICONfromQImage(const QImage &image) const
+{
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    auto p = QPixmap::fromImage(image);
+    return QtWin::toHICON(p);
+#else
+    return image.toHICON();
+#endif
+}
+
+void TaskBarNotifier::Private::doFlashTaskbarIcon()
+{
+    const int count = 0;
+    const int speed = 0;
+
+    FLASHWINFO fi;
+    fi.cbSize    = sizeof(FLASHWINFO);
+    fi.hwnd      = hwnd_;
+    fi.dwFlags   = (urgent_) ? FLASHW_ALL | FLASHW_TIMER : FLASHW_STOP;
+    fi.uCount    = count;
+    fi.dwTimeout = speed;
+    FlashWindowEx(&fi);
+}
+#endif
+
+TaskBarNotifier::TaskBarNotifier(QWidget *parent, const QString &desktopfile) :
+    count_(0), icon_(nullptr), active_(false)
+{
+    d = new Private(this);
+#ifdef USE_DBUS
+    Q_UNUSED(parent)
+    d->setDesktopPath(desktopfile);
+#elif defined(Q_OS_WINDOWS)
+    Q_UNUSED(desktopfile);
+    HWND hwnd = reinterpret_cast<HWND>(parent->winId());
+    d->setParentHWND(hwnd);
+    d->setDevicePixelRatio(parent->devicePixelRatio());
+#endif
+}
+
+TaskBarNotifier::~TaskBarNotifier()
+{
+    if (icon_)
+        delete icon_;
```

Comment:
you can use std::unique_ptr instead.
this will make sure
1) icon is deleted in constructor
2) icon is deleted on consequent call to setIconCountCaption